### DISCUSSION TITLE
Porting code to other MCUs

### DISCUSSION THREAD

**Discussion by manuelbl** [manuelbl/usb-pd-arduino#2] [Ideas]
What will it take to port the code to further MCUs?

For STM32 MCUs, it should be straight-forward. The existing code for USB PD processing (timer and DMA in input capture mode, processing of captured values, SPI for transmission etc.) should contain all the building blocks for implementing all other STM32 MCUs. The main challenge can be to figure out a working combination of pins, timers and timer channels as the STM32 is quite restrictive in this regard.

For other MCU families, it looks different. For certain ones I'm skeptical it can be done at all as they are missing crucial components such as some sort of input capture or a peripheral suitable for generating the CC signal.

What's your take? What would be a suitable approach for the ESP32 or the Raspberry Pico?

**Comment by chatelao** [comment]
I start to get used to the RP2040 and it's PIO module. I bet my *** it shouldn't be too difficult to implement the required Rx/Tx state machine.

**Comment by manuelbl** [comment]
The Raspi Pico is most likely capable to deal with USB PD. And the PIO peripheral can indeed be helpful.

Do you already have a setup in mind, e.g. using external comparators to level shift the USB PD signals to 3.3V CMOS levels and the PIO peripheral for measuring the duration of the high and low state of the input signal or even fully decoding it, and for encoding the output signal?
